Checkpoint Review

alex-magana opened the pull request:
#### What does this PR do?

Create a flight booking application
#### Screenshots (if appropriate)

<img width="1680" alt="screen shot 2016-10-19 at 8 29 41 am" src="https://cloud.githubusercontent.com/assets/20963836/19506687/52e01ef0-95d6-11e6-8b62-1c2b6a1b227f.png">

alex-magana committed:
Implement flight booking feature

Use nested forms to implement flight booking.

alex-magana committed:
Add side nav

Add a side navigation to the home page for ease of use.

alex-magana committed:
Add booking search and user authentication

Add booking search feature and user authentication

alex-magana committed:
Implement testing and mailing feature

alex-magana commented on app/controllers/bookings_controller.rb:
```diff
@@ -32,20 +44,63 @@ def create
+      @booking = Booking.where("bookings.booking_code = ?", params[:booking_code]).first
+      render 'edit'
+    else
+      @booking = Booking.select("bookings.*, flights.flight_code, flights.airline_code, flights.departure_date, flights.arrival_date")
```
Move the logic to the model.

ChigboIO replied:
You said it yourself. Who am I to say otherwise? 😆

alex-magana commented on app/controllers/bookings_controller.rb:
```diff
@@ -9,17 +11,27 @@ def show
   end
 
   def new
-    @flight = Flight.find(5)
```
Method should not exceed 8 lines

Habu-Kagumba commented on app/assets/javascripts/materialize.js:
```diff
@@ -1,6 +1,11 @@
 $(document).ready(function() {
   $(".button-collapse").sideNav();
 
+  $("#menu").click(function() {
```
Line 2 already takes care of the sidenav. I don't think you need to add another event listener on the same button.

Habu-Kagumba commented on app/controllers/bookings_controller.rb:
```diff
@@ -9,17 +11,27 @@ def show
     @booking = Booking.new
+    @booking.user_id = current_user
     @booking.flight = @flight
+    @booking.booking_code = SecureRandom.hex(4).upcase
```
This looks like it belongs in the model.

ChigboIO replied:
One way to do what @Habu-Kagumba suggests will be to either override the `new` method (after all, you won't have a booking without user and flight) or a different method in the model that takes `current_user` and `flight` objects and make and return new **booking** with those

ChigboIO replied:
You can also use a `before_create` to set  that _booking_code_ and not bother about passing it to the view and back to the controller

Habu-Kagumba commented on app/controllers/bookings_controller.rb:
```diff
@@ -9,17 +11,27 @@ def show
+    if current_user
+      @booking.email = current_user.email
+    end
+    if current_user
```
This is a second `if` statement with the same condition (with `if` at `line 19`).

ChigboIO replied:
Yeah. You should combine them if you can

Habu-Kagumba commented on app/controllers/bookings_controller.rb:
```diff
@@ -32,20 +44,63 @@ def create
+  end
+
+  def past_bookings
+    @bookings = Booking.select("bookings.*, flights.flight_code, flights.airline_code, flights.departure_date, flights.arrival_date")
```
The DB specific logic seems out of place here. This should be in the model maybe as a `scope`.

ChigboIO replied:
Yes. Please use `scope` for this and the likes. For when to use `scope`, see this http://www.justinweiss.com/articles/should-you-use-scopes-or-class-methods/

Habu-Kagumba commented on app/controllers/flights_controller.rb:
```diff
@@ -47,10 +23,23 @@ def update
+  def search_flights
+    @params = flight_params
+
+    @flights = Flight.select("flights.*, airlines.airline_name")
```
Should be e.g. a scope in the model.

Habu-Kagumba commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  end
+
+  def user_authenticate
+    @user = User.where("email = ? and password = ?", user_params[:email], user_params[:password]).first
```
The `where` clause seems like model logic that should alternatively live in the model. This appears in several places across the controllers.

Habu-Kagumba commented on app/mailers/wings_mailer.rb:
```diff
@@ -0,0 +1,6 @@
+class WingsMailer < ApplicationMailer
+  def send_email(email)
+    binding.pry
```
You should remove debug code.

ChigboIO replied:
😱 😆

alex-magana committed:
Migrate db from sqlite to postgres

alex-magana committed:
Add sqlite to postgresql migration files

alex-magana committed:
Set database config

Update database config for development, test and production.

alex-magana committed:
Implement passing model tests and UI improvemments

ChigboIO commented on Gemfile:
```diff
@@ -5,6 +5,8 @@ source 'https://rubygems.org'
 gem 'rails', '~> 5.0.0', '>= 5.0.0.1'
 # Use sqlite3 as the database for Active Record
 gem 'sqlite3'
```
The `sqlite3` gem should be moved to a `development, test` block if this app will be hosted (or is hosted) on Heroku

ChigboIO commented on app/assets/javascripts/materialize.js:
```diff
@@ -1,6 +1,16 @@
+      // menuWidth: 300,
+      // closeOnClick: true
+    // }
+  // );
```
Remove the commented lines since they aren't doing anything

ChigboIO replied:
Ideally, any query that is not `find` should go to the model as a `scope` or `class method` with a better descriptive name. Ideally! 😄

ChigboIO commented on app/controllers/bookings_controller.rb:
```diff
@@ -9,17 +11,26 @@ def show
+    if current_user
+      @booking.user_id = current_user.id
+    else
+      @booking.user_id = 2
```
Any reason you're hard-coding user_id 2 here?

ChigboIO commented on app/controllers/bookings_controller.rb:
```diff
@@ -1,5 +1,7 @@
 class BookingsController < ApplicationController
   before_action :set_booking, only: [:show, :edit, :update, :destroy]
+  before_action :new_booking_params, only: [:new]
+  before_action :booking_params, only: [:confirm_booking, :create]
```
**Strong params** are not action filters, they are just used to protect model attributes against mass assignment. There is no point making it a `before_action` Check [this](http://edgeapi.rubyonrails.org/classes/ActionController/StrongParameters.html) for clarity.

ChigboIO commented on app/controllers/users_controller.rb:
```diff
@@ -0,0 +1,68 @@
+
+  def create
+    @user = User.new(user_params)
+    @user.role_id = 2
```
This line is not very descriptive as to what the value `2` mean. Consider using `enum` enhance this. Read up here http://www.justinweiss.com/articles/creating-easy-readable-attributes-with-activerecord-enums/

ChigboIO commented on config/application.yml:
```diff
@@ -0,0 +1,7 @@
+password: 'fc205a5e07a8ce848b0d87917314e0d9'
+
+gmail_username: 'alex.magana@andela.com'
+gmail_password: 'asd!@#6353aMgx'
```
This file should be git-ignored. I assume you're using **figaro** so next time, run:

``` shell
$ bundle exec figaro install
```

This creates the file and adds it to `.gitignore` file

Since you are already tracking this file, you can run:

``` shell
$ git rm --cached config/application.yml
```

to remove it from git

Habu-Kagumba replied:
😱

ChigboIO commented on db/migrate/20161007052847_create_users.rb:
```diff
@@ -0,0 +1,14 @@
+      t.string :last_name
+      t.string :email
+      t.string :password
+      t.belongs_to :role, foreign_key: true
```
Nice! I didn't know of this alias `belongs_to`. I only use `reference`. 👍

ChigboIO reviewed (commented):
Nice work @andela-amagana Awesome flow and code style 👍

alex-magana committed:
Implement mailing and eliminate sturcture errors in views

alex-magana committed:
Update Gemfile to set sqlite for development and test

Set sqlite to be used for development and test environments.

alex-magana committed:
Generate seed data

alex-magana committed:
Rearrange models in seed file

Organize dummy data to ensure adherence to table associations.

alex-magana committed:
Add production environment email configuration

alex-magana committed:
Delete irrelevant dummy data

Delete the db_export directory
Trim the seed file

alex-magana committed:
Add figaro to manage environment variables

alex-magana committed:
Enable testing js response format

Disable 'protect_from_forgery' for the test environment.

alex-magana committed:
Consume scope defined in the model

Move logic from the controller.

alex-magana committed:
Eliminate new line between methods

alex-magana committed:
Implement scope to accommodate custom DML

alex-magana committed:
Rename find_flight to search_flights

alex-magana committed:
Rectify syntax errors

Include tags to properly access variable values

alex-magana committed:
Add tests for bookings controller

alex-magana committed:
Add tests for flights controller

alex-magana committed:
Remove 'user_id' to allow dynamic assignment

alex-magana committed:
Change new-user to class

alex-magana committed:
Move the logic to model user

Use action new for sign up

alex-magana committed:
Add scope to query data

alex-magana committed:
Add 'Reserve flight' link

alex-magana committed:
Use form_tag to define the form

alex-magana committed:
Render signup page

alex-magana committed:
Remove signup path

alex-magana committed:
Define variables flight and booking

alex-magana committed:
Use a partial to render sign up

alex-magana committed:
Define tests for controller user

alex-magana committed:
Add coveralls to the project

alex-magana committed:
Add coveralls badge in README.md

alex-magana committed:
Add code climate to the repository

alex-magana committed:
Update coverage files and add travis config

alex-magana committed:
Add selenium to support feature testing

alex-magana committed:
Remove unused code

alex-magana committed:
Eliminate white background in submit buttons

alex-magana committed:
Correct syntax in scope where clause

alex-magana committed:
Remove pry invocation

alex-magana committed:
Add classes to button submit

alex-magana committed:
Add id attribute to each form element

alex-magana committed:
Add div to display flash messages

alex-magana committed:
Ensure spec examples pass

alex-magana committed:
Correct error in create example spec

Eliminate wrong model count increment.
Please enter the commit message for your changes. Lines starting

alex-magana committed:
Include user_id in factory

alex-magana committed:
Add newline between distinct config blocks

alex-magana committed:
Add search_flight and add_booking_details in module Helpers

alex-magana committed:
Add feature tests for registered and anonymous users

alex-magana committed:
Rename travis file

alex-magana committed:
Change db initialization commands

alex-magana committed:
Correct class referencing

alex-magana committed:
Update test coverage

alex-magana committed:
Include seed command

alex-magana committed:
Specify Firefox version

alex-magana committed:
Rectify selenium driver error

alex-magana committed:
Rectify selenium web driver error

alex-magana committed:
Specify example to find text "Sign In" in root path

Locate "Sign In" in header bar

alex-magana committed:
Add sleep to allow time to find element

alex-magana committed:
Eliminate check for 'Sign In'

alex-magana committed:
Increase sleep duration to facilitate element search

alex-magana committed:
Concentrate inspection in main-nav

alex-magana committed:
Set default wait time to improve inspection

alex-magana committed:
Remove within block definition

alex-magana committed:
Set default wait time for all examples

alex-magana committed:
Correct errors in controllers

Rectify incorrect redirection and database querying.

alex-magana committed:
Initialize materialize elements on turbolink load

alex-magana committed:
Change header and footer height

alex-magana committed:
Position search-booking-col div in the middle of row

alex-magana committed:
Add margin and background styling for search results

alex-magana committed:
Remove box-shadow for the user view fields

alex-magana committed:
Use lambda for multiline scopes

alex-magana committed:
Remove faulty waves-effect and waves-light classes

alex-magana committed:
Center the search field in the view

alex-magana committed:
Remove faulty waves-effect waves-light class

alex-magana committed:
Set padding for the search bar

alex-magana committed:
Remove faulty waves-effect waves-light class

alex-magana committed:
Render partial _form_edit

alex-magana committed:
Remove faulty waves-effect waves-light class

alex-magana committed:
Add top padding for the row div

Specify padding to position the form below the nav bar

alex-magana committed:
Render partial _form_show

alex-magana committed:
Persist passenger quantity specified by the user

alex-magana committed:
Resize column size for input fields

alex-magana committed:
Resize div search and available-container

alex-magana committed:
Persist passenger quantity specified in search view

alex-magana committed:
Disable turbolinks

alex-magana committed:
Add padding for the search criteria fields

alex-magana committed:
Remove unnecessary comments

alex-magana committed:
Implement sticky footer

alex-magana committed:
Center flight search page fields

alex-magana committed:
Add method to provide current user's id

alex-magana committed:
Refactor bookings_controller

Eliminate repetition and adhere to the style guide

alex-magana committed:
Refactor flights_controller

Adhere to the style guide

alex-magana committed:
Refactor users_controller

Adhere to the style guide

alex-magana committed:
Replace scope with class methods

alex-magana committed:
Remove height and overflow specification

alex-magana committed:
Invoke renamed method

alex-magana committed:
Specify association

alex-magana committed:
Specify association and replace scope with class method

alex-magana committed:
Refactor to enforce style guide

alex-magana committed:
Correct the error message wording

alex-magana committed:
Replace past_bookings with index

Use index controller action and view in place of past_bookings

alex-magana committed:
Render edit link for a registered user

alex-magana committed:
Remove div id option-fields

alex-magana committed:
Replace past_bookings with index

Use index controller action and view in place of past_bookings

alex-magana committed:
Remove div id option-fields

alex-magana committed:
Remove div id option fields and disable turbolinks

alex-magana committed:
Implement responsiveness

alex-magana committed:
Remove div id option-fields

alex-magana committed:
Redirect 'Past bookings' to bookings_path

alex-magana committed:
Remove wrong error message and implement responsiveness

alex-magana committed:
Remove unnecessary views

alex-magana committed:
Remove faulty matrialize waves classes

alex-magana committed:
Remove purged routes

alex-magana committed:
Refactor controller specs

alex-magana committed:
Refactor feature specs

alex-magana committed:
Refactor model specs

alex-magana committed:
Configure DatabaseCleaner

alex-magana committed:
Refactor helper methods

alex-magana committed:
Update gems, coverage and schema

alex-magana committed:
Define partial for editing bookings

alex-magana committed:
Define partial for viewing bookings

alex-magana committed:
Change db structure

alex-magana committed:
Rectify failing specs

alex-magana committed:
Rectify failing flights_controller spec

alex-magana committed:
Increase sleep time for feature specs

alex-magana committed:
Refactor tests for circleci build

alex-magana committed:
Update circleci repo token

alex-magana committed:
Remove travis config file

alex-magana committed:
Update coveralls config

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update README.md

alex-magana committed:
Update controller and feature specs

alex-magana committed:
Change view layout to improve UI/UX

alex-magana committed:
Refactor controllers

alex-magana committed:
Refactor styling and event listeners

alex-magana committed:
Add .DS_Store to .gitignore

alex-magana committed:
Change error message rendering in log in

alex-magana committed:
Enforce foreign key constraint on passengers model

alex-magana committed:
Render a different view for anonymous users

alex-magana committed:
Add codeclimate test reporter

alex-magana committed:
Update README.md

alex-magana committed:
Edit format of the booking details email

alex-magana committed:
Add specs for model class methods

alex-magana committed:
Validate presence of passengers in booking instance

alex-magana committed:
Eliminate sleep time in feature specs

alex-magana committed:
Rectify failing tests

alex-magana committed:
Abstract retrieving user with role walk_in

alex-magana committed:
Integrate thoughtbot configuration

alex-magana committed:
Refactor controllers

alex-magana committed:
Refactor views

alex-magana committed:
Refactor models

alex-magana committed:
Expose sessions routes

alex-magana committed:
Update tables and add relationships

alex-magana committed:
Remove fields from flight seed data

alex-magana committed:
Refactor tests

alex-magana committed:
Add styling for sessions pages

alex-magana committed:
Remove source_airport and destination_airport from flights

alex-magana committed:
Rename sessions routes

alex-magana committed:
Use .class_method notation for model specs
